pkg/apis/storagemigration/types.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storagemigration
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.29
+
+// StorageVersionMigration represents a migration of stored data to the latest
+// storage version.
+type StorageVersionMigration struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the migration.
+	Spec StorageVersionMigrationSpec
+	// Status of the migration.
+	// +optional
+	Status StorageVersionMigrationStatus
+}
+
+// Spec of the storage version migration.
+type StorageVersionMigrationSpec struct {
+	// The resource that is being migrated from. The migrator sends requests to
+	// the endpoint serving the resource.
+	// Immutable.
+	OriginResource metav1.GroupVersionResource
+	// The storage version hash to avoid races.
```

Comment:
Can we expand the documentation of this field to more clearly say what it is and why it is here?

Reply:
Also, it's not clear totally clear to me why this is in Spec, can we put some kind of note giving the background or justification for that?

Reply:
@jpbetz I'm not entirely sure if we need `StorageVersionHash`. Can we remove it? https://github.com/kubernetes/kubernetes/pull/121123#discussion_r1361359667